File: rocket-chat-core/src/main/java/chat/rocket/core/interactors/CanCreateRoomInteractor.java
package chat.rocket.core.interactors;

import io.reactivex.Flowable;
import io.reactivex.Single;

import chat.rocket.core.repositories.UserRepository;

public class CanCreateRoomInteractor {

  private final UserRepository userRepository;
  private final SessionInteractor sessionInteractor;

  public CanCreateRoomInteractor(UserRepository userRepository,
                                 SessionInteractor sessionInteractor) {
    this.userRepository = userRepository;
    this.sessionInteractor = sessionInteractor;
  }

  public Single<Boolean> canCreate(String roomId) {
    return Flowable.zip(
        userRepository.getCurrent(),
        sessionInteractor.getDefault(),
        Flowable.just(roomId),
        (user, session, room) -> user != null && session != null && room != null
    )
        .first(false);
  }
}

File: rocket-chat-core/src/main/java/chat/rocket/core/interactors/CanCreateRoomInteractor.kt
package chat.rocket.core.interactors

import chat.rocket.core.models.Session
import chat.rocket.core.models.User
import io.reactivex.Flowable
import io.reactivex.Single

import chat.rocket.core.repositories.UserRepository
import com.hadisatrio.optional.Optional
import io.reactivex.functions.Function3

class CanCreateRoomInteractor(private val userRepository: UserRepository,
                              private val sessionInteractor: SessionInteractor) {

    fun canCreate(roomId: String): Single<Boolean> {
        return Flowable.zip<Optional<User>, Optional<Session>, String, Boolean>(
                userRepository.getCurrent(),
                sessionInteractor.getDefault(),
                Flowable.just(roomId),
                Function3 { user, session, room -> user.isPresent && session.isPresent && room != null }
        )
                .first(false)
    }
}
